src/main/java/com/google/sps/servlets/ProjectServlet.java

```diff
@@ -0,0 +1,266 @@
+      return;
+    }
+
+    // Mode is a required parameter
+    boolean isCreateMode = DataUtils.parseMode(request, response);
+
+    String uEmail = userService.getCurrentUser().getEmail();
+    String projId = request.getParameter("proj-id");
+
+    // Will be either a new entity for creation or an existing entity for
+    // updating
+    Entity projEntity = new Entity(DataUtils.PROJECT);
+
+    if (!isCreateMode) {
+      // Must be owner to update
+      projEntity = DataUtils.getProjectEntity(projId, uEmail, false, false);
+
+      // Delete overrides all other updates
+      boolean delete = Boolean.parseBoolean(request.getParameter("delete"));
+      if (delete) {
+        datastore.delete(projEntity.getKey());
+        response.sendRedirect("/"); // placeholder: should redirect to projects gallery
+        return;
+      }
+    }
+
+    // Last modified timestamp
+    String now = Instant.now().toString();
+    projEntity.setProperty("utc", now);
+
+    String projName = request.getParameter("proj-name");
+
+    // Set the project name if provided
+    // If creating and nothing provided, set name to Untitled-{current UTC time}
+    if (!DataUtils.isEmptyParameter(projName)) {
+      projEntity.setProperty("name", projName);
+    } else {
+      if (isCreateMode) {
+        projEntity.setProperty("name", "Untitled-" + now);
+      }
+      // Don't do anything if updating and no name provided
+    }
+
+    // Default visibility to private (only owners and editors can view)
+    String visibility = request.getParameter("visibility");
+    if (DataUtils.isEmptyParameter(visibility) && isCreateMode) {
+      visibility = DataUtils.PRIVATE;
+    }
+    if (!DataUtils.isEmptyParameter(visibility) &&
+        (visibility.toLowerCase().equals(DataUtils.PUBLIC) ||
+         visibility.toLowerCase().equals(DataUtils.PRIVATE))) {
+      projEntity.setProperty("visibility", visibility);
+    }
+
+    String ownersString = request.getParameter("owners");
+    String editorsString = request.getParameter("editors");
+
+    // If creating:
+    //      owners are current User and any people in parameter
+    //      if no one provided, current User is the only owner
+    // if updating:
+    //      owners are current User and any people in parameter
+    //      if no one provided, change nothing
+    if (isCreateMode || !DataUtils.isEmptyParameter(ownersString)) {
+      ArrayList<String> listOwnerEmails = new ArrayList<String>();
+      if (!DataUtils.isEmptyParameter(ownersString)) {
+        listOwnerEmails = DataUtils.parseCommaList(ownersString);
+      }
+      listOwnerEmails.add(uEmail);
+      projEntity.setIndexedProperty(
+          "owners", DataUtils.withDuplicatesRemoved(listOwnerEmails));
+    }
+
+    // If anything provided for editors, overwrite current editors
+    if (!DataUtils.isEmptyParameter(editorsString)) {
+      ArrayList<String> listEditorEmails =
+          DataUtils.parseCommaList(editorsString);
```

Comment:
What if the user enters the list with semi-colons seperating, or just spaces?

Reply:
It would not work with the current implementation. Unsure how semicolons vs commas vs spaces would be differentiated, especially if spaces could be used within tags.